lib.systems: rust: Make rustcTargetSpec the primary entrypoint

dramforever opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

**For reviewers**: If there's a Rust derivation under some platform that

- Builds with current Nixpkgs
- Has eval changes at all

I consider this a bug in this PR. This is my goal.

- - -

This is what I would call "Project `rustcTarget = cargoShortTarget`". Well, that's the middle step, where the broken and the working coexist. Eventually we should really clean up `*Platform.{rust,rustc}`.

I want to make custom Rust target work in Nixpkgs, with something as "simple" as:

```
import <nixpkgs> {
  crossSystem = {
    config = "mips64el-unknown-linux-gnuabi64";
    gcc = { # Obviously you need to configure the C compiler too
      arch = "mips3";
      float = "hard";
      abi = "64";
    };
    rust.rustcTargetSpec = "${./rust}/mips64el_mips3-unknown-linux-gnuabi64.json";
  };
}
```

The reason I believe that `rust.platform` couldn't work is because we cannot create directories with `builtins.toFile`. I do not believe that the old `toFile` approach is worth preserving - if you're doing custom Rust targets, you most likely have a JSON target specification anyway. It would be easier to use that.

Currently, I have some basic std-using stuff working like microfetch. I also got stc-ng working once upon a time (https://github.com/dramforever/nixos-loongson2f) using this approach to specify basenames.

More observations:

`rustcTarget` is broken. I'm pretty sure nobody knows what it is supposed to be. Every single use in Nixpkgs should be replaced with either `rustcTargetSpec` or `cargoShortTarget`. So, whatever it was supposed to be for, it clearly didn't happen. If it were to come up again somehow, a new name would be better.

This custom target stuff is also *really* undertested, because the only user of custom Rust target in Nixpkgs is `rust-hypervisor-firmware` and that's it.

I also don't know why #95542 says std doesn't work for custom targets...

TODO:

- [X] doc: Explain `"${./path/to/dir}"/target-spec.json"` better (Better enough for review)
- [X] Test `lib` changes for breakages (No changes in `lib.mapAttrs (_: lib.systems.elaborate) lib.systems.examples`)
- [X] Test rust-hypervisor-firmware (No eval changes, tested with `NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1`)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [X] aarch64-linux (cross compiling `microfetch` to ["mips64el_mips3-unknown-linux-gnuabi64"](https://github.com/dramforever/nixos-loongson2f/blob/8b37597eb8ec08b71cb37d2d4c7fd97eab4eccf7/rust/mips64el_mips3-unknown-linux-gnuabi64.json), and running under `qemu-mips64el`, also confirmed that the target has been applied)
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [X] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dramforever committed:
doc/rust: Use rustcTargetSpec everywhere

Cargo uses the basename of a JSON target specification in various places
to refer to targets, cc-rs parses the basename to grab target
information. There may be other examples where the basename is relevant.

Instead of fighting these existing conventions, let's just recommend
users provide a rustcTargetSpec, either as a standard name or as a
custom JSON file with the right basename

dramforever commented:
The `lib` commit contains some rather large code movements and indentation changes. Consider reading the diff with whitespace changes ignored.

dramforever commented:
The `lib` commit contains some rather large code movements and indentation changes. Consider reading the diff with whitespace changes ignored.

alyssais reviewed (approved):
Makes sense to me.  The diff is a bit difficult to read indeed (even with whitespace disabled) so I'm not totally confident in all the subtleties but I didn't notice anything immediately wrong.

@Ericson2314, do you have any feelings on this part?

> The reason I believe that rust.platform couldn't work is because we cannot create directories with builtins.toFile. I do not believe that the old toFile approach is worth preserving - if you're doing custom Rust targets, you most likely have a JSON target specification anyway. It would be easier to use that.

dramforever commented on lib/systems/default.nix:
```diff
@@ -488,29 +496,30 @@ let
+                else
+                  "${cpu_}-${vendor_}-${kernel.name}${optionalString (abi.name != "unknown") "-${abi.name}"}";
             in
             # TODO: deprecate args.rustc in favour of args.rust after 23.05 is EOL.
```
Maybe. I think later. :)

dramforever reviewed (commented):
Oops, typo...

Can't easily work on Git rn, posting this here so I don't forget.

dramforever committed:
lib.systems: Refactor rust.platform generation

In preparation for a future commit, refactor rust.platform somewhat.

No functional changes intended. Most of the diff here is whitespace
changes.

dramforever committed:
lib.systems: rust: Make rustcTargetSpec the primary entrypoint

Make rustcTargetSpec the primary entrypoint for setting a custom target,
and wire up all the other stuff so they are hopefully as working and as
broken as before.

In particular, to specify a custom target, the user now just specifies
rust.rustcTargetSpec. rust.platform and rust.cargoShortTarget are
populated from rust.rustcTargetSpec now. In addition, rust.rustcTarget
defaults to rust.cargoShortTarget. (rust.rustcTarget and
rust.cargoShortTarget really should always be the same, but I think we
can deal with that later).

This allows the user to more easily control the basename of
rust.rustcTargetSpec by passing e.g.
"${./rust}/mips64el_mips3-unknown-linux-gnuabi64.json", which allows
cc-rs and in turn std to work.

dramforever commented:
While rebasing, I also did some commit splitting, so now it should be somewhat readable if you review each commit individually.

alyssais reviewed (approved):
Looks fine, but at some point we really need to start cleaning up the legacy stuff. It's getting out of hand.

dramforever commented on lib/systems/default.nix:
```diff
@@ -502,28 +508,28 @@ let
             );
 
+          # Do not use rustcTarget. Use rustcTargetSpec or cargoShortTarget.
+          # FIXME: Remove all in-tree usages
```
and deprecate.

I was gonna make a list of things to deprecate but `grep TODO lib/systems/default.nix` is good enough. And the next step is slapping warnings on everything and fix everything in-tree.
